bundled/tool/server.py

```diff
@@ -223,6 +257,108 @@ def _get_settings_by_document(document: workspace.Document | None):
 # *****************************************************
 # Internal execution APIs.
 # *****************************************************
+CODE_TO_SUMMARY: dict[str, str] | None = None
+
+
+def _reset_code_to_summary():
+    global CODE_TO_SUMMARY
+    CODE_TO_SUMMARY = None
+
+
+def _init_code_to_summary(document: workspace.Document) -> None:
+    global CODE_TO_SUMMARY
+
+    if CODE_TO_SUMMARY is not None:
+        return
+
+    log_to_output("Initializing CODE_TO_SUMMARY mapping")
+
+    if str(document.uri).startswith("vscode-notebook-cell"):
+        # Skip notebook cells
+        return None
+
+    if utils.is_stdlib_file(document.path):
+        # Skip standard library python files.
+        return None
+
+    # deep copy here to prevent accidentally updating global settings.
+    settings = copy.deepcopy(_get_settings_by_document(document))
+
+    code_workspace = settings["workspaceFS"]
+    cwd = settings["workspaceFS"]
+
+    use_path = False
+    use_rpc = False
+    if settings["path"]:
+        # 'path' setting takes priority over everything.
+        use_path = True
+        argv = settings["path"]
+    elif settings["interpreter"] and not utils.is_current_interpreter(settings["interpreter"][0]):
+        # If there is a different interpreter set use JSON-RPC to the subprocess
+        # running under that interpreter.
+        argv = [TOOL_MODULE]
+        use_rpc = True
+    elif settings["importStrategy"] == "useBundled":
+        # If we're loading from the bundle, use the absolute path.
+        argv = [
+            os.fspath(pathlib.Path(__file__).parent.parent / "libs" / "bin" / TOOL_MODULE),
+        ]
+    else:
+        # If the interpreter is same as the interpreter running this process then run
+        # as module.
+        argv = [os.path.join(sysconfig.get_path("scripts"), TOOL_MODULE)]
+
+    argv += ["--show-rules", "--format", "json", "."]
```

Comment:
@charliermarsh What do you think about this CLI option?

Reply:
Maybe more generally useful would be to add `ruff --explain F841`, similar to `clippy --explain`. Then it could print an explanation, or format it as JSON, and we could call out to that command when the user hovers over the check (rather than building up the map in advance). Would that work?

Reply:
> Maybe more generally useful would be to add ruff --explain F841

Agreed. Yes, this also would work.